src/initialization.jl

```diff
@@ -30,3 +36,155 @@ struct OverrideInitData{IProb, UIProb, IProbMap, IProbPmap}
 end
+
+"""
+    get_initial_values(prob, valp, f, alg, isinplace; kwargs...)
+
+Return the initial `u0` and `p` for the given SciMLProblem and initialization algorithm,
+and a boolean indicating whether the initialization process was successful. Keyword
+arguments to this function are dependent on the initialization algorithm. `prob` is only
+required for dispatching. `valp` refers the appropriate data structure from which the
+current state and parameter values should be obtained. `valp` is a non-timeseries value
+provider as defined by SymbolicIndexingInterface.jl. `f` is the SciMLFunction for the
+problem. `alg` is the initialization algorithm to use. `isinplace` is either `Val{true}`
+if `valp` and the SciMLFunction are inplace, and `Val{false}` otherwise.
+"""
+function get_initial_values end
+
+struct CheckInitFailureError <: Exception
+    normresid::Any
+    abstol::Any
+end
+
+function Base.showerror(io::IO, e::CheckInitFailureError)
+    print(io,
+        "CheckInit specified but initialization not satisfied. normresid = $(e.normresid) > abstol = $(e.abstol)")
+end
+
+struct OverrideInitMissingAlgorithm <: Exception end
+
+function Base.showerror(io::IO, e::OverrideInitMissingAlgorithm)
+    print(io,
+        "OverrideInit specified but no NonlinearSolve.jl algorithm provided. Provide an algorithm via the `nlsolve_alg` keyword argument to `get_initial_values`.")
+end
+
+"""
+Utility function to evaluate the RHS of the ODE, using the integrator's `tmp_cache` if
+it is in-place or simply calling the function if not.
+"""
+function _evaluate_f_ode(integrator, f, isinplace::Val{true}, args...)
+    tmp = first(get_tmp_cache(integrator))
+    f(tmp, args...)
+    return tmp
+end
+
+function _evaluate_f_ode(integrator, f, isinplace::Val{false}, args...)
+    return f(args...)
+end
+
+"""
+    $(TYPEDSIGNATURES)
+
+A utility function equivalent to `Base.vec` but also handles `Number` and
+`AbstractSciMLScalarOperator`.
+"""
+_vec(v) = vec(v)
+_vec(v::Number) = v
+_vec(v::SciMLOperators.AbstractSciMLScalarOperator) = v
+_vec(v::AbstractVector) = v
+
+"""
+    $(TYPEDSIGNATURES)
+
+Check if the algebraic constraints are satisfied, and error if they aren't. Returns
+the `u0` and `p` as-is, and is always successful if it returns. Valid only for
+`ODEProblem` and `DAEProblem`. Requires a `DEIntegrator` as its second argument.
+"""
+function get_initial_values(prob::ODEProblem, integrator, f, alg::CheckInit,
+        isinplace::Union{Val{true}, Val{false}}; kwargs...)
+    u0 = state_values(integrator)
+    p = parameter_values(integrator)
+    t = current_time(integrator)
+    M = f.mass_matrix
+
+    algebraic_vars = [all(iszero, x) for x in eachcol(M)]
+    algebraic_eqs = [all(iszero, x) for x in eachrow(M)]
+    (iszero(algebraic_vars) || iszero(algebraic_eqs)) && return
+    update_coefficients!(M, u0, p, t)
+    tmp = _evaluate_f_ode(integrator, f, isinplace, u0, p, t)
+    tmp .= ArrayInterface.restructure(tmp, algebraic_eqs .* _vec(tmp))
```

Comment:
This allocates

Reply:
Somehow changing this to `tmp = algebraic_eqs .* _vec(tmp)` allocates _more_

Reply:
At least, according to `@allocated`/`@allocations`